File: frontend/lib/models/caixa.dart
enum TipoTransacao {
  receita,
  despesa;

  String get valor => this == TipoTransacao.receita ? 'receita' : 'despesa';
  String get rotulo => this == TipoTransacao.receita ? 'Receita' : 'Despesa';

  static TipoTransacao fromValor(String v) => v == 'receita' ? TipoTransacao.receita : TipoTransacao.despesa;
}

enum MetodoPagamento {
  numerario,
  mpesa,
  emola;

  String get valor {
    switch (this) {
      case MetodoPagamento.numerario:
        return 'numerario';
      case MetodoPagamento.mpesa:
        return 'mpesa';
      case MetodoPagamento.emola:
        return 'emola';
    }
  }

  String get rotulo {
    switch (this) {
      case MetodoPagamento.numerario:
        return 'Numerário';
      case MetodoPagamento.mpesa:
        return 'M-Pesa';
      case MetodoPagamento.emola:
        return 'E-Mola';
    }
  }

  static MetodoPagamento? fromValor(String? v) {
    if (v == null) return null;
    return MetodoPagamento.values.firstWhere((m) => m.valor == v, orElse: () => MetodoPagamento.numerario);
  }
}

/// Categorias padrão usadas pelo ecrã de Inscrições e Renovações — os
/// valores exatos que o backend guarda e mostra.
class CategoriaCaixa {
  static const inscricao = 'Inscrição';
  static const renovacao = 'Renovação';
  static const fichaCatecumeno = 'Ficha do Catecúmeno';
}

class CaixaTransacao {
  final String id;
  final TipoTransacao tipo;
  final String categoria;
  final double valor;
  final MetodoPagamento? metodoPagamento;
  final String? catequisandoId;
  final String? catequisandoNome;
  final String? faseId;
  final String? faseNome;
  final int? anoLetivo;
  final String? descricao;
  final DateTime data;
  final String registadoPorNome;

  CaixaTransacao({
    required this.id,
    required this.tipo,
    required this.categoria,
    required this.valor,
    this.metodoPagamento,
    this.catequisandoId,
    this.catequisandoNome,
    this.faseId,
    this.faseNome,
    this.anoLetivo,
    this.descricao,
    required this.data,
    required this.registadoPorNome,
  });

  factory CaixaTransacao.fromJson(Map<String, dynamic> json) => CaixaTransacao(
        id: json['id'] as String,
        tipo: TipoTransacao.fromValor(json['tipo'] as String),
        categoria: json['categoria'] as String,
        valor: (json['valor'] as num).toDouble(),
        metodoPagamento: MetodoPagamento.fromValor(json['metodo_pagamento'] as String?),
        catequisandoId: json['catequisando_id'] as String?,
        catequisandoNome: json['catequisando_nome'] as String?,
        faseId: json['fase_id'] as String?,
        faseNome: json['fase_nome'] as String?,
        anoLetivo: json['ano_letivo'] as int?,
        descricao: json['descricao'] as String?,
        data: DateTime.parse(json['data'] as String),
        registadoPorNome: json['registado_por_nome'] as String,
      );
}

class ResumoCaixa {
  final double totalReceitas;
  final double totalDespesas;
  final double saldo;

  ResumoCaixa({required this.totalReceitas, required this.totalDespesas, required this.saldo});

  factory ResumoCaixa.fromJson(Map<String, dynamic> json) => ResumoCaixa(
        totalReceitas: (json['total_receitas'] as num).toDouble(),
        totalDespesas: (json['total_despesas'] as num).toDouble(),
        saldo: (json['saldo'] as num).toDouble(),
      );
}

File: backend/app/models/caixa.py
from datetime import date, datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class TipoTransacao(str, Enum):
    RECEITA = "receita"
    DESPESA = "despesa"


class MetodoPagamento(str, Enum):
    NUMERARIO = "numerario"
    MPESA = "mpesa"
    EMOLA = "emola"


# Categorias sugeridas para receitas ligadas a um catequisando — usadas pelo
# ecrã de Inscrições e Renovações. O campo em si é texto livre (permite
# outras categorias, ex. despesas), mas o frontend usa sempre estes valores
# para as 3 taxas específicas da catequese.
CATEGORIA_INSCRICAO = "Inscrição"
CATEGORIA_RENOVACAO = "Renovação"
CATEGORIA_FICHA_CATECUMENO = "Ficha do Catecúmeno"

# Categorias que representam a matrícula do catequisando num ano letivo —
# a fase é guardada nestas transações, e é daqui que se deriva o histórico
# de "em que fase estava X no ano Y" (ver GET /catequisandos/{id}/historico).
CATEGORIAS_MATRICULA = {CATEGORIA_INSCRICAO, CATEGORIA_RENOVACAO}


class CaixaTransacaoCreate(BaseModel):
    tipo: TipoTransacao
    categoria: str = Field(..., min_length=2, max_length=60)
    valor: float = Field(..., gt=0)
    metodo_pagamento: Optional[MetodoPagamento] = None
    catequisando_id: Optional[str] = None
    fase_id: Optional[str] = Field(None, description="Fase a que a inscrição/renovação se refere")
    ano_letivo: Optional[int] = Field(None, description="Se omitido, usa o ano letivo corrente")
    descricao: Optional[str] = Field(None, max_length=300)
    data: date


class CaixaTransacaoUpdate(BaseModel):
    tipo: Optional[TipoTransacao] = None
    categoria: Optional[str] = Field(None, min_length=2, max_length=60)
    valor: Optional[float] = Field(None, gt=0)
    metodo_pagamento: Optional[MetodoPagamento] = None
    catequisando_id: Optional[str] = None
    fase_id: Optional[str] = None
    ano_letivo: Optional[int] = None
    descricao: Optional[str] = Field(None, max_length=300)
    data: Optional[date] = None


class CaixaTransacaoOut(BaseModel):
    id: str
    tipo: TipoTransacao
    categoria: str
    valor: float
    metodo_pagamento: Optional[MetodoPagamento] = None
    catequisando_id: Optional[str] = None
    catequisando_nome: Optional[str] = None
    fase_id: Optional[str] = None
    fase_nome: Optional[str] = None
    ano_letivo: Optional[int] = None
    descricao: Optional[str] = None
    data: date
    registado_por_nome: str
    criado_em: datetime


class ResumoCaixa(BaseModel):
    total_receitas: float
    total_despesas: float
    saldo: float
